mcp23017_driver/mcp23017.h

```diff
@@ -0,0 +1,351 @@
+    uint8_t device_address; ///I2C address of the MCP23017 device.
+    uint8_t i2c_timeout; ///Timeout value for I2C communication in milliseconds
+} MCP23017_t;
+
+/**
+ * @typedef
+ * @brief Specifies whether to refer to the PORTA or PORTB register.
+ */
+typedef enum {
+    port_a,
+    port_b,
+} port_t;
+
+typedef enum {
+    logic_high = 1,
+    logic_low  = 0,
+} logic_pin_level_t;
+
+typedef enum {
+    io_direction_input  = 1,
+    io_direction_output = 0,
+} io_direction_t;
+
+typedef enum {
+    input_polarity_inverted = 1,
+    input_polarity_normal   = 0,
+} input_polarity_mode_t;
+
+typedef enum {
+    interrupt_enabled  = 1,
+    interrupt_disabled = 0,
+} interrupt_status_t;
+
+typedef enum {
+    interrupt_compare_with_ref_value  = 1,
+    interrupt_compare_with_prev_value = 0,
+} interrupt_compare_mode_t;
+
+typedef enum {
+    pull_up_resistor_enabled = 1,
+    pull_up_resistor_disable = 0,
+} pull_up_resistor_status_t;
+
+typedef enum {
+    flag_high = 1,
+    flag_low  = 0,
+} flag_status_t;
+
+/**
+ * @paragraph Simple
+ * 
+ * The following function are for simplified use of the device.
+*/
+
+/**
+ * @brief Get current configuration from device.
+ * @param hdev Pointer to the MCP23017 device structure.
+ * @param config Pointer to the MCP23017 configuration structure.
+ * 
+ * @return HAL status from communication with the device via I2C.
+ */
+HAL_StatusTypeDef get_config(MCP23017_t* hdev, MCP23017_configuration_t* config);
+
+/**
+ * @brief Set configuration to device.
+ */
+HAL_StatusTypeDef set_config(MCP23017_t* hdev, MCP23017_configuration_t config);
+
+/**
+ * @brief reset configuration to the default. refer to the default_configuration struct. 
+*/
+HAL_StatusTypeDef reset_config(MCP23017_t* hdev);
+
+/**
+ * @brief Initializes the MCP23017 using the specified configuration.
+ *
+ * @param hdev Pointer to the MCP23017 device structure.
+ * @param hi2c Pointer to the I2C handle used for communication.
+ * @param device_address I2C address of the MCP23017 device.
+ * @param i2c_timeout Timeout value for I2C communication in milliseconds.
+ * @param config Configuration settings for the MCP23017.
+ * 
+ * @return HAL status from communication with the device via I2C.
+ */
+HAL_StatusTypeDef init(MCP23017_t* hdev, I2C_HandleTypeDef* hi2c, uint8_t device_address, uint8_t i2c_timeout, MCP23017_configuration_t config);
+
+/**
+ * @brief Not fully implemented.
+ */
+//HAL_StatusTypeDef init_default(MCP23017_t* hdev, I2C_HandleTypeDef* hi2c);
+
+/**
+ * @brief
+ * 
+ * https://electronics.stackexchange.com/questions/325916/mcp23017-detecting-state-of-iocon-bank-bit-after-mcu-reset
+ * 
+ * You cannot determine the state of IOCON.BANK without hardware resetting it.
+ * I implemented a routine to get it into a known value without resetting it.
+ * This function will bring the device always to IOCON.BANK = 0 without modifying anything else.
+ * 
+*/
+HAL_StatusTypeDef reset_bank_config(MCP23017_t* hdev);
```

Comment:
I don't know if it is called or not in the init, but avoid messing with this register. You neither touch it nor read it and it will be in his default state and everybody is happy.